src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -0,0 +1,208 @@
+  /// alpha and beta factors set to 1.
+  /// The constraints on rank were taken from the torch implementation check
+  /// Input A must be of rank 2 (input)
+  /// Input B must be of rank 2 (weights)
+  /// Input C must be of rank 1 (bias)
+  static bool checkLegalTosaFCOp(ONNXGemmOp op, OpAdaptor adaptor) {
+    Value A = op.A();
+    Value B = op.B();
+    Value C = op.C();
+
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+
+    // If C is present, it can only be of rank 1
+    if (C.getType().isa<RankedTensorType>() &&
+        C.getType().cast<RankedTensorType>().getRank() != 1) {
+      return false;
+    }
+    // Input tensor must be of rank 2
+    // Weights must also be of rank 2
+    if (AType.getRank() != 2 || BType.getRank() != 2) {
+      return false;
+    }
+    // Both alpha and beta must be 1
+    if ((adaptor.alpha().convertToFloat() != 1.0F) ||
+        (adaptor.beta().convertToFloat() != 1.0F)) {
+      return false;
+    }
+    // Only Transpose B must be enabled
+    if (adaptor.transA() != 0 || adaptor.transB() != 1) {
+      return false;
+    }
+    return true;
+  }
+  LogicalResult matchAndRewrite(ONNXGemmOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+    Value A = op.A();
+    Value B = op.B();
+    Value C = op.C();
+    int64_t transA = adaptor.transA();
+    int64_t transB = adaptor.transB();
+    FloatAttr alpha = adaptor.alphaAttr();
+    FloatAttr beta = adaptor.betaAttr();
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+    auto shapeA = A.getType().dyn_cast<ShapedType>().getShape();
+    auto shapeB = B.getType().dyn_cast<ShapedType>().getShape();
+    auto resultType = getTypeConverter()->convertType(op.getResult().getType()).cast<TensorType>();
+    
+    // C is optional, if it's not there, we need to be aware of it for later computations
+    bool isCPresent = C.getType().isa<TensorType>();
+    // If it's not present, ONNX creates a NoValue operator. Remove it.
+    if (!isCPresent) {
+      rewriter.eraseOp(C.getDefiningOp());
+    }
+
+    // If legal, create a FullyConnected operator instead
+    if (checkLegalTosaFCOp(op, adaptor)) {
+      return rewriteToTosaFC(op, adaptor, rewriter);
+    }
+    
+    // ONNX gives 2d matrix as input and expect a 2d output. TOSA expects everything to be 3D. As such, there
+    // is a need to add reshapes operators before and after we do any computation. We set the batch as 1 as it
+    // is unknown.
+    llvm::SmallVector<int64_t> newShapeA{1};
+    llvm::SmallVector<int64_t> newShapeB{1};
+    newShapeA.append({shapeA[0], shapeA[1]});
+    newShapeB.append({shapeB[0], shapeB[1]});
+
+    A = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op->getLoc(),
+        UnrankedTensorType::get(AType.getElementType()), A,
```

Comment:
While `UnrankedTensorType` is allowed, I read https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Tosa/IR/TosaTypesBase.td#L123 as suggesting it is preferable to use a ranked tensor if you can do so. I'm not an expert here, so if you can point out why unranked would be preferred I'd be happy to learn.

Reply:
I just think it's easier because I don't have to compute the shape myself, and CreateOpAndInfer already runs shape inference to resize the tensors as wanted

Reply:
`RankedTensorType::get({-1, -1, -1}, AType.getElementType())` is an alternative, but you at least check that the tensor has the right rank.

Reply:
Both are fine by me, if you think it's better I don't have a strong opinion